drv/medusa-seq-server/src/main.rs

```diff
@@ -0,0 +1,357 @@
+                //
+                // Whether or not the PHY oscillator was found to be operating
+                // nominally is recorded in the front IO board controller. Look
+                // up what this value is to determine if a power reset of the
+                // front IO board is needed.
+                match front_io_board.phy().osc_state()? {
+                    PhyOscState::Bad => {
+                        // The PHY was attempted to be initialized but its
+                        // oscillator was deemed not functional. Unfortunately
+                        // the only course of action is to power cycle the
+                        // entire front IO board, so do so now.
+                        self.power_control.v12_qsfp_out.set_enable(false);
+                        ringbuf_entry!(Trace::FrontIOBoardPowerEnable(false));
+
+                        // Wait some cool down period to allow caps to bleed off
+                        // etc.
+                        userlib::hl::sleep_for(1000);
+                    }
+                    PhyOscState::Good => {
+                        // The PHY was initialized properly before and its
+                        // oscillator declared operating nominally. Assume this
+                        // has not changed and only a reset the PHY itself is
+                        // desired.
+                        front_io_board
+                            .phy()
+                            .set_phy_power_enabled(false)
+                            .map_err(MedusaError::from)?;
+                        ringbuf_entry!(Trace::FrontIOBoardPhyPowerEnable(
+                            false
+                        ));
+
+                        userlib::hl::sleep_for(10);
+                    }
+                    PhyOscState::Unknown => {
+                        // Do nothing (yet) since the oscillator state is
+                        // unknown.
+                    }
+                }
+            }
+        }
+
+        // Run preinit to check HSC status.
+        self.front_io_board_preinit()?;
+
+        if let Some(front_io_board) = self.front_io_board.as_mut() {
+            // At this point the front IO board has either not yet been
+            // initalized or may have been power cycled and should be
+            // initialized.
+            if !front_io_board.initialized() {
+                front_io_board.init()?;
+            }
+
+            // The PHY is still powered down. Request the sequencer to power up
+            // and wait for it to be ready.
+            front_io_board.phy().set_phy_power_enabled(true)?;
+            ringbuf_entry!(Trace::FrontIOBoardPhyPowerEnable(true));
+
+            while !front_io_board.phy().powered_up_and_ready()? {
+                userlib::hl::sleep_for(20);
+            }
+
+            Ok(())
+        } else {
+            Err(MedusaError::NoFrontIOBoard)
+        }
```

Comment:
style nit, take it or leave it: personally, i might consider writing this using `ok_or` rather than `if let`, so that the returned error is closer to the condition in which we return the error, and the code reads like a straight line:
```suggestion
        let front_io_board = self
            .front_io_board
            .as_mut()
            .ok_or(MedusaError::NoFrontIOBoard)?;

        // At this point the front IO board has either not yet been
        // initialized or may have been power cycled and should be
        // initialized.
        if !front_io_board.initialized() {
            front_io_board.init()?;
        }

        // The PHY is still powered down. Request the sequencer to power up
        // and wait for it to be ready.
        front_io_board.phy().set_phy_power_enabled(true)?;
        ringbuf_entry!(Trace::FrontIOBoardPhyPowerEnable(true));

        while !front_io_board.phy().powered_up_and_ready()? {
            userlib::hl::sleep_for(20);
        }

        Ok(())
```

Reply:
Stylistically I think I've tended to see the `if let Some` pattern more in Hubris, but I like your suggestion better!